torch/distributed/pipelining/stage.py

```diff
@@ -695,6 +695,13 @@ def backward_one_chunk(
                 self.grads_input = grads_input
                 # Save a placeholder for the dw_runner
                 self.dw_runner[bwd_chunk_id] = lambda: None
+
+        if self.is_last:
+            # stage_output is no longer used in the last stage for backward and only needed
+            # to return to the user in merge_output_chunks, therefore
+            # this should be detached to release autograd graph context and free memory earlier
+            for t in stage_output:
+                t.detach_()
```

Comment:
i'm confused about the difference in the 2 PRs.  The first PR detaches 'stage_output' in stage_backward_input, which iiuc runs inside this function (backward_one_chunk) but above this point, in the case of non-full-backward.  If this is the last stage, wouldn't we have already run stage_backward_input and detached stage_outputs by the time we get here?  or are there multiple copies of 'stage_output' and we have to detach them both?

Reply:
For the last stage `backward_one_chunk` operates on the loss. The dependencies look like:

`rest_of_autograd_graph` -> `stage_output` -> `loss`

The "`stage_outputs`" detached in the first PR was actually the loss for the last stage (i guess it should be renamed to be `stage_output_or_loss`. So we also need to also detach the `stage_output` in the last stage to allow the memory to be freed

Reply:
gotcha. yea, rename for clarity.  and maybe even put that simple illustration in to the comment `rest_of_autograd_graph -> stage_output -> loss`